src/Shared/IInternable.cs

```diff
@@ -0,0 +1,346 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Text;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build
+{
+    #region IInternable
+    /// <summary>
+    /// Define the methods needed to intern something.
+    /// </summary>
+    internal interface IInternable
+    {
+        /// <summary>
+        /// The length of the target.
+        /// </summary>
+        int Length { get; }
+
+        /// <summary>
+        /// Indexer into the target. Presumed to be fast.
```

Comment:
Since as pointed out below it's not necessarily fast, it might be best to remove this from the interface and replace with `bool EqualsOrdinal(IInternable)` and on the concrete types have specialized overloads for the other concrete types.

Reply:
Hm, I see you need a hashcode as well. Maybe string builders should hash only their first N chars?

Reply:
Yes, having our `StringBuilder` wrapper implements the interface throws a wrench into what would otherwise be a straightforward refactor. Will be addressed in a separate PR.

Reply:
> To make the hash a bit better, maybe first n/3 characters, last n/3 characters, and middle n/3 characters?

Whatever you do should be based on the perf characteristics of StringBuilders. It's an implementation detail, but unlikely to change: if the SB is fragmented, the SB object you hold contains the buffer of the last fragment, and a pointer to the next previous. Indexing into the SB requires walking along the chain - there is no "index" and there is no "cached last index". So perhaps I should have said the last N characters; jumping around in the SB is what is costly. It's essentially a character buffer designed to be cheaply modified at the very end, then turned into a contiguous buffer or string and thrown away.